### ISSUE TITLE
OnnxRuntime Exception

### ISSUE THREAD

**Issue by amrutha95** [microsoft/onnxruntime#19727]
### Describe the issue

I've been trying to get OCR working in a windows app, specifically using this model
[microsoft/trocr-base-handwritten · Hugging Face](https://huggingface.co/microsoft/trocr-base-handwritten). I converted this to an onnx model but I'm running into issues trying to run inference in a WinUI app. This is the error:

Microsoft.ML.OnnxRuntime.OnnxRuntimeException: '[ErrorCode:RuntimeException] Non-zero status code returned while running Add node. Name:'/embeddings/Add' Status Message: C:\__w\1\s\onnxruntime\onnxruntime\core/providers/cpu/math/element_wise_ops.h:560 onnxruntime::BroadcastIterator::Append axis == 1 || axis == largest was false. Attempting to broadcast an axis by a dimension other than 1. 577 by 1222
'

I have confirmed that the onnx is valid, I used onnxruntime in Python and I was able to run inference without any error. I'm wondering if my input is not formatted correctly but there's no way to tell from this error message. 

### To reproduce

Link to download the models : https://microsoft-my.sharepoint-df.com/:u:/p/amsrin/ETofgNRkjPhLmtVynWwtAk0B_T7cOpexxjFFJPR3wkXvIA?e=33HFHQ
Please note that since this is an encoder decoder model, the onnx files are split.

```
 var image = Image.Load<Rgb24>(filePath);

 DenseTensor<float> tensorImage = new(new[] { 1, 3, image.Height, image.Width });

 image.ProcessPixelRows(accessor =>
 {
     for (int y = 0; y < accessor.Height; y++)
     {
         Span<Rgb24> pixelSpan = accessor.GetRowSpan(y);
         for (int x = 0; x < accessor.Width; x++)
         {
             tensorImage[0, 0, y, x] = pixelSpan[x].R;
             tensorImage[0, 1, y, x] = pixelSpan[x].G;
             tensorImage[0, 2, y, x] = pixelSpan[x].B;
         }
     }
 });

 using var inputOrtValue = OrtValue.CreateTensorValueFromMemory(OrtMemoryInfo.DefaultInstance,
      tensorImage.Buffer, new long[] { 1, 3, image.Height, image.Width });

 var inputs = new Dictionary<string, OrtValue>
 {
     { "pixel_values", inputOrtValue }
 };


 using var session = new InferenceSession("path-to-model.onnx");
 using var runOptions = new RunOptions();
 runOptions.LogVerbosityLevel = 1;
 try
 {
     using IDisposableReadOnlyCollection<OrtValue> results = session.Run(runOptions, inputs, session.OutputNames);
 }
catch(Exception e) {
//Execution lands up in this catch block due to the exception
}
```

### Platform

Windows

### OS Version

23H2

### ONNX Runtime Installation

Released Package

### ONNX Runtime Version or Commit ID

1.17.1

### ONNX Runtime API

C#

### Architecture

X64

### Execution Provider

Default CPU

**Comment by pranavsharma** [comment]
Did the exact same model work when using the Python bindings? Are the ORT versions the same?

**Comment by amrutha95** [comment]
Yes the same model worked. I just checked, in python I was running 1.15.1 and in .NET it was 1.17.1

**Comment by djrecipe** [comment]
Since your onnx isn't publicly available, I can only speculate. What method did you take in converting in to the onnx format?

**Comment by amrutha95** [comment]
Hi, sorry I didn't realize the link was not public. Try [this one](https://1drv.ms/f/s!AsxozzAU-hnooRdKGmhHznT62BtY?e=QTqzur)

**Comment by amrutha95** [comment]
As for converting it to the onnx format, I used the optimum cli

**Comment by yuslepukhin** [comment]
The problem occurs when an Add node attempts to Add the output of a Concat node /embeddings/Concat_1 with an initializer. Add is a elementwise operation which requires the same shape or the shape that is broadcastable. But the dimensions can not be reconciled:
The out from Concat is of shape [1, 1222, 768] and the initializer 
embeddings.position_embeddings is of shape [1,577,768]

The model is not valid.

**Comment by zhufuguo** [comment]
> The problem occurs when an Add node attempts to Add the output of a Concat node /embeddings/Concat_1 with an initializer. Add is a elementwise operation which requires the same shape or the shape that is broadcastable. But the dimensions can not be reconciled: The out from Concat is of shape [1, 1222, 768] and the initializer embeddings.position_embeddings is of shape [1,577,768]
> 
> The model is not valid.

**I encounter a problem that may caused by the same bug.  Below are my debuging settings.**
1.	Windows 10 platform，using cpu for debugging.
2.	Python + ORT and Pytorch get the same correct result.
3.	C++ & ORT cause runtime exception in ORT (runtime crash).
4.	Python ORT and C++ ORT are manually built from ORT Git repo with version tag v1.17.1. 
5.	The data received by ORT internal function below are binary the same between python and C++ calling (write to file and diff).
Status InferenceSession::Run(const RunOptions& run_options, gsl::span<const std::string> feed_names, gsl::span<const OrtValue> feeds, gsl::span<const std::string> output_names, std::vector<OrtValue>* p_fetches, const std::vector<OrtDevice>* p_fetches_device_info)

**From the above we can conclude that:**
1.	The exported onnx model is valid.
2.	There are diff between python and c++ calling of ORT.
**The problem is:**
1.	C++ and Python share the core component of ORT, but with their own API encapsulation codes. Does this bug caused by possible different configurations in API code?
2.	The Dynamic library loading mechanism between C++ and python may cause global variable or static variable initialization diff?  If so, the bug may locate in ORT core part.

**Power shell commend line running C++ test demo:**
[E:onnxruntime:, sequential_executor.cc:514 onnxruntime::ExecuteKernel] Non-zero status code returned while running FusedConv node. Name:'/dec/resblocks.4/convs1.1/Conv'

**Visual studio run without debug C++ test demo:** 
[E:onnxruntime:, sequential_executor.cc:514 onnxruntime::ExecuteKernel] Non-zero status code returned while running Add node. Name:'/enc_p/Add_2' Status Message: C:\workspace\onnxruntime\onnxruntime\core/providers/cpu/math/element_wise_ops.h:560 onnxruntime::BroadcastIterator::Append axis == 1 || axis == largest was false. Attempting to broadcast an axis by a dimension other than 1. 5 by 43

**Comment by yuslepukhin** [comment]
Python output using provided test image.

> Loading D:/dev/data/gh_issue_19727_OCR/encoder_model_opt.onnx model took 3.5742084980010986 seconds.
['pixel_values']
['last_hidden_state']
input shape:  (1, 3, 605, 531)
========== D:/dev/data/gh_issue_19727_OCR/encoder_model_opt.onnx failed: [ONNXRuntimeError] : 6 : RUNTIME_EXCEPTION : Non-zero status code returned while running Add node. Name:'/embeddings/Add' Status Message: D:\dev\ort_main\onnxruntime\core/providers/cpu/math/element_wise_ops.h:560 onnxruntime::BroadcastIterator::Append axis == 1 || axis == largest was false. Attempting to broadcast an axis by a dimension other than 1. 577 by 1222
Stacktrace:
D:\dev\ort_main\onnxruntime\core\providers\cpu\math\element_wise_ops.h(677): onnxruntime_pybind11_state!onnxruntime::Broadcaster::Broadcaster+0x95B
D:\dev\ort_main\onnxruntime\core\providers\cpu\math\element_wise_ops.h(783): onnxruntime_pybind11_state!onnxruntime::InputBroadcaster::InputBroadcaster+0x20

**Comment by zhufuguo** [comment]
> The problem occurs when an Add node attempts to Add the output of a Concat node /embeddings/Concat_1 with an initializer. Add is a elementwise operation which requires the same shape or the shape that is broadcastable. But the dimensions can not be reconciled: The out from Concat is of shape [1, 1222, 768] and the initializer embeddings.position_embeddings is of shape [1,577,768]
> 
> The model is not valid.

Hi,  the c++ code, python code, onnx model and test data to reproduce my problem are attached below：

https://drive.google.com/file/d/1pYXdqBh_c7gBL5YspKTcbhrKx_mJTj8V/view?usp=drive_link

**Comment by yuslepukhin** [comment]
@amrutha95 I just reproduced the problem with Python 1.1.5.1, which is no suriprise.

**Comment by zhufuguo** [comment]
After a week of debugging, my problem was resolved. 
It was my improper calling of the C++version API. The variable of type Ort:: Env needs to remain valid throughout the entire program lifecycle.  
 
```
 env = Ort::Env(ORT_LOGGING_LEVEL_WARNING, "debug");
 session_options = Ort::SessionOptions();
```
A model with fixed input and output dimensions may not make mistakes, while a model with variable input and output dimensions is likely to make mistakes. 
Thank you all for your information.

**Comment by djrecipe** [comment]
@zhufuguo do you mind still providing access to the reproduction code to brkfstmnchr@gmail.com? I'm interested in taking a look at this
